### PR TITLE
Add useful handlers for admin

### PR DESCRIPTION
**Pull request by decebals**
I added small route handlers that are useful in my opinion for an admin.

```java
GET("/ping", new PingHandler());
GET("/threadDump", new ThreadDumpHandler());
GET("/metrics", new MetricsHandler());
GET("/healthCheck", new HealthCheckHandler(healthCheckRegistry));
GET("/settings", new SettingsHandler());
GET("/sysInfo", new SystemInfoHandler());
```

### CODE REVIEW

File: pippo-metrics-parent/pippo-metrics/src/main/java/ro/pippo/metrics/ThreadDumpHandler.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    @Override
+    public void handle(RouteContext routeContext) {
+        Response response = routeContext.getResponse().noCache().text();
+
+        if (threadDump != null) {
+            OutputStream output = response.getOutputStream();
+            threadDump.dump(output);
+        } else {
+            response.internalError().send("Sorry your runtime environment does not allow to dump threads");
```
Comment (mhagnumdw) [inline pippo-metrics-parent/pippo-metrics/src/main/java/ro/pippo/metrics/ThreadDumpHandler.java]: `internalError()` return 500, not 501. JavaDoc says

> If the thread dump is not supported then the status code returned is 501 (not implemented).

File: pippo-core/src/main/java/ro/pippo/core/route/SystemInfoHandler.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        Map<String, String> props = new HashMap<>();
+        for (String name : System.getProperties().stringPropertyNames()) {
+            props.put(name, System.getProperty(name));
+        }
+
+        Map<String, String> env = System.getenv();
+
+        Response response = routeContext.getResponse().noCache().text();
+
+        try (BufferedWriter writer = new BufferedWriter(response.getWriter())) {
```
Comment (mhagnumdw) [inline pippo-core/src/main/java/ro/pippo/core/route/SystemInfoHandler.java]: What do you think of differentiating in output what is system property and what is environment variable?

**Comment by mhagnumdw** [comment]
> GET("/metrics", new MetricsHandler());

I did not find the class MetricsHandler.

Overall the idea seems to be cool!

File: pippo-metrics-parent/pippo-metrics/src/main/java/ro/pippo/metrics/ThreadDumpHandler.java (hunk 1 above)
Comment (decebals) [reply pippo-metrics-parent/pippo-metrics/src/main/java/ro/pippo/metrics/ThreadDumpHandler.java]: I made the correction.

**Comment by decebals** [comment]
In the future I want to add query parameter to these handlers (where is applicable) to filter some information. For example I want to retrieve only the system properties from `SystemInfoHandler` so, I will use "sysprops" as parameter.

Another idea is to adapt somehow the content type of the responses. Now these handlers return text only (maybe I want an `html` table or a `json`). From this reason I introduce the protected method, to have a possibility to "format" the content/result. It's only an idea and I am not sure that is valuable.

File: pippo-core/src/main/java/ro/pippo/core/route/SystemInfoHandler.java (hunk 2 above)
Comment (mhagnumdw) [reply pippo-core/src/main/java/ro/pippo/core/route/SystemInfoHandler.java]: This approach is good.

**Comment by decebals** [comment]
> I did not find the class MetricsHandler.

This handler is not prod ready. The idea is that it's relative complex to convert a [MetricRegistry](https://metrics.dropwizard.io/3.1.0/apidocs/com/codahale/metrics/MetricRegistry.html) to plain text.
We can use [MetricsModule](https://metrics.dropwizard.io/3.1.0/apidocs/com/codahale/metrics/json/MetricsModule.html) to convert in `json` but this solution comes with a dependency to [metrics-json](https://mvnrepository.com/artifact/com.codahale.metrics/metrics-json). To convert metrics to json using metrics-json library can be, in the end, the correct solution (or at least, an acceptable solution).

**Comment by decebals** [comment]
I added `MetricsHandler`.

See below some screenshots of this admin handlers:

**SettingsHandler**
image

**SystemInfoHandler**
image

**ThreadDumpHandler**
image

**MetricsHandler**
image

**Comment by decebals** [comment]
See below how you can add a new `HealthCheck` implementation (`DatabaseHealthCheck`):

```java
import com.codahale.metrics.health.HealthCheck;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.sql.DataSource;

import java.sql.Connection;
import java.sql.DatabaseMetaData;
import java.sql.SQLException;
import java.sql.Statement;

public class DatabaseHealthCheck extends HealthCheck {

    private static final Logger log = LoggerFactory.getLogger(DatabaseHealthCheck.class);

    private final DataSource dataSource;
    private final String healthCheckSql;

    private final String healthyMessage;
    private final String unHealthyMessage;

    public DatabaseHealthCheck(DataSource dataSource, String healthCheckSql) throws SQLException {
        this.dataSource = dataSource;
        this.healthCheckSql = healthCheckSql;

        DatabaseMetaData metaData = dataSource.getConnection().getMetaData();
        String url = metaData.getURL();
        String userName = metaData.getUserName();

        healthyMessage = String.format("Database connection successful (host = %s, username %s", url, userName);
        unHealthyMessage = String.format("Unable to connect to database (host = %s, username %s'", url, userName);
    }

    @Override
    protected Result check() {
        Result result;

        try (Connection connection = dataSource.getConnection(); Statement statement = connection.createStatement()) {
            boolean ok = statement.execute(healthCheckSql);
            result = ok ? Result.healthy(healthyMessage) : Result.unhealthy(unHealthyMessage);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            result = Result.unhealthy(unHealthyMessage);
        }

        log.debug(result.getMessage());

        return result;
    }

}
```

and now you must register the new health check implementation:
```java
import com.codahale.metrics.health.HealthCheckRegistry;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class MyApplication extends Application {

    private static final Logger log = LoggerFactory.getLogger(MyApplication.class);

    private final HealthCheckRegistry healthCheckRegistry;

    public MyApplication() {
        healthCheckRegistry = new HealthCheckRegistry();
        try {
            healthCheckRegistry.register("database", new DatabaseHealthCheck(DbUtils.getDataSource(), "SELECT 1 FROM dual"));
        } catch (SQLException e) {
            log.error(e.getMessage(), e);
        }
    }

    @Override
    protected void onInit() {
        GET("/healthCheck", new HealthCheckHandler(healthCheckRegistry));
    }

}
```

**Pull request merged**